File: play_signal/generate_signal.py
import numpy as np
from scipy.signal import resample

SAMPLE_RATE = 10000
MIN_FREQ = 50
MAX_FREQ = 500

# Load calibration Coeffs
Coeffs = []
with open('play_signal/Coeff.txt') as file:
    for line in file:
        Coeffs.append(10000 / float(line.strip()))
m = np.mean(Coeffs)
Coeffs = np.array(Coeffs) / m
assert len(Coeffs) == (MAX_FREQ - MIN_FREQ + 1), "Coeff.txt length mismatch"


def map_frequency(value, min_value, max_value, logscale=False):
    if logscale:
        log_min = np.log10(MIN_FREQ)
        log_max = np.log10(MAX_FREQ)
        log_freq = log_min + (log_max - log_min) * ((value - min_value) / (max_value - min_value))
        return 10 ** log_freq
    else:
        return MIN_FREQ + (MAX_FREQ - MIN_FREQ) * ((value - min_value) / (max_value - min_value))


def linterp_index(arr, perc):
    index = perc * (len(arr) - 1)
    lower = int(np.floor(index))
    upper = int(np.ceil(index))
    weight = index - lower
    return arr[lower] * (1 - weight) + arr[upper] * weight


def generate_signal(vib_amp, vib_freq, TOTAL_DURATION_SEC, duration=None, logscale=False):
    if duration is None:
        duration = TOTAL_DURATION_SEC
    elif duration > TOTAL_DURATION_SEC:
        raise ValueError(f"duration must be <= {TOTAL_DURATION_SEC} seconds")

    total_samples = int(TOTAL_DURATION_SEC * SAMPLE_RATE)
    num_output_samples = int(duration * SAMPLE_RATE)
    vib_signal = np.zeros(num_output_samples, dtype=np.float32)

    phase_acc = 0
    for i in range(num_output_samples):
        perc = i / total_samples
        s_freq = linterp_index(vib_freq, perc)
        s_amp = linterp_index(vib_amp, perc)

        current_freq = map_frequency(s_freq, 0, 1, logscale=logscale)
        phase_delta = 2 * np.pi * current_freq / SAMPLE_RATE

        phase_acc += phase_delta
        vib_signal[i] = s_amp * np.sin(phase_acc) * Coeffs[int(current_freq) - MIN_FREQ]

    return vib_signal, duration


def generate_signal_with_thermal(vib_amp, vib_freq, thermal_amp, TOTAL_DURATION_SEC, duration=None, logscale=False, thermal_rate=10000):
    vib_signal, actual_duration = generate_signal(vib_amp, vib_freq, TOTAL_DURATION_SEC, duration, logscale)

    # vib_signal의 길이에 맞춰 thermal_amp를 interpolation
    vib_len = len(vib_signal)
    thermal_full = np.interp(
        np.linspace(0, len(thermal_amp) - 1, vib_len),
        np.arange(len(thermal_amp)),
        thermal_amp
    )

    # downsample to thermal_rate (예: 100Hz)
    thermal_sampled = resample(thermal_full, int(actual_duration * thermal_rate))

    return vib_signal, thermal_sampled

File: static/generate-signal.mjs
//@ts-check
import { DrawCanvasElement } from "./drawcanvas.mjs";
import { notnull } from "./util.mjs";

export const SAMPLE_RATE = 10000;
export const MIN_FREQ = 50;
export const MAX_FREQ = 500;
export const MIN_THERMAL = -6;
export const MAX_THERMAL = 6;


/** @type {DrawCanvasElement} */
const vib_amp_env_dcanvas = notnull(document.querySelector("#drawvibaenv"));
/** @type {DrawCanvasElement} */
const vib_freq_env_dcanvas = notnull(document.querySelector("#drawvibfenv"));
/** @type {DrawCanvasElement} */
const thr_amp_env_dcanvas = notnull(document.querySelector("#drawthraenv"));
/** @type {HTMLInputElement} */
const duration_input = notnull(document.querySelector("#signalduration"));

/**
 * Maps a given frequency value to the MIN_FREQ - MAX_FREQ range.
 * @param {number} value - The original frequency value.
 * @param {number} minValue - The minimum value in the dataset.
 * @param {number} maxValue - The maximum value in the dataset.
 * @returns {number} - The mapped frequency in the range [MIN_FREQ, MAX_FREQ].
 */
function map_frequency(value, minValue, maxValue) {
    return linear_map(value, minValue, maxValue, MIN_FREQ, MAX_FREQ);
}
/**
 * Maps a given thermal value to the MIN_THERMAL to MAX_THERMAL range.
 * @param {number} value - The original thermal value.
 * @param {number} minValue - The minimum value in the dataset.
 * @param {number} maxValue - The maximum value in the dataset.
 * @returns {number} - The mapped thermal value in the range [MIN_THERMAL, MAX_THERMAL].
 */
export function map_thermal(value, minValue, maxValue) {
    return linear_map(value, minValue, maxValue, MIN_THERMAL, MAX_THERMAL);
}
/**
 * Inversely maps a given thermal value from the MIN_THERMAL to MAX_THERMAL range to the original range.
 * @param {number} value - The original thermal value.
 * @param {number} minValue - The minimum value in the dataset.
 * @param {number} maxValue - The maximum value in the dataset.
 * @returns {number} - The inversely mapped thermal value in the range [minValue, maxValue].
 */
export function inv_map_thermal(value, minValue, maxValue) {
    return linear_map(value, MIN_THERMAL, MAX_THERMAL, minValue, maxValue);
}

/**
 * Maps a given value from one range to another.
 * @param {number} value - The original value.
 * @param {number} orig_min - The minimum value in the original range.
 * @param {number} orig_max - The maximum value in the original range.
 * @param {number} new_min - The minimum value in the new range.
 * @param {number} new_max - The maximum value in the new range.
 * @returns {number} - The mapped value in the new range.
 */
function linear_map(value, orig_min, orig_max, new_min, new_max) {
    return new_min + (new_max - new_min) * ((value - orig_min) / (orig_max - orig_min));
}

function linterp_index(arr, perc) {
    const index = perc * (arr.length - 1);
    const lower = Math.floor(index);
    const upper = Math.ceil(index);
    const weight = index - lower;
    const value = arr[lower] * (1 - weight) + arr[upper] * weight;
    return value;
}


/**
 * Generate a signal based on the current envelope values.
 * @param {boolean} [force_full_duration=false] - If true, the full duration of the signal is generated regardless of the active duration.
 * @returns {{ vib_signal: Float32Array, thermal_signal: Float32Array, duration: number }}
 */
export function generate_signal(force_full_duration = false) {
    const vib_amp = vib_amp_env_dcanvas.get_samples();
    const vib_freq = vib_freq_env_dcanvas.get_samples();
    const thr_amp = thr_amp_env_dcanvas.get_input_samples();

    const current_total_duration = vib_amp_env_dcanvas.get_total_duration();
    if (current_total_duration != vib_freq_env_dcanvas.get_total_duration() || current_total_duration != thr_amp_env_dcanvas.get_total_duration()) {
        throw new Error("All cavnases must have the same total duration");
    }
    const current_active_duration = vib_amp_env_dcanvas.get_active_duration();
    if (current_active_duration != vib_freq_env_dcanvas.get_active_duration() || current_active_duration != thr_amp_env_dcanvas.get_active_duration()) {
        throw new Error("All cavnases must have the same active duration");
    }
    const total_sample_duration = current_total_duration;
    const active_duration = force_full_duration ? current_total_duration : current_active_duration;
    const active_perc = active_duration / total_sample_duration;
    const num_output_samples = Math.floor(active_duration * SAMPLE_RATE);
    const vib_signal = new Float32Array(num_output_samples);

    let phase_acc = 0;
    for (let i = 0; i < num_output_samples; i++) {
        const perc = i / num_output_samples;
        const lperc = perc * active_perc;

        const s_freq = linterp_index(vib_freq, lperc);
        const s_amp = linterp_index(vib_amp, lperc);

        const current_freq = map_frequency(s_freq, 0, 1);
        const phase_delta = 2 * Math.PI * current_freq / SAMPLE_RATE;

        phase_acc += phase_delta;
        vib_signal[i] = s_amp * Math.sin(phase_acc);
    }

    const thermal_signal = new Float32Array(num_output_samples);
    for (let i = 0; i < num_output_samples; i++) {
        const perc = i / num_output_samples;
        const lperc = perc * active_perc;

        const s_amp = linterp_index(thr_amp, lperc);

        // Map [0, 1] to [-6, 6]
        const current_s_amp = map_thermal(s_amp, 0, 1);
        thermal_signal[i] = current_s_amp
    }

    return { vib_signal, thermal_signal, duration: active_duration };
}